### PR TITLE
Feature/custom tmx loader

### PR DESCRIPTION
**Pull request by BWalusiak**

### CODE REVIEW

File: core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+
+    public static TileMap createTileMap(World world, String path) throws IOException, MissingLayersException, ParserConfigurationException, SAXException {
+
+        Node sourceMap = loadXmlMap(path);
+        Element mapElement = (Element) sourceMap;
+        NodeList tilesetSource = mapElement.getElementsByTagName("tileset");
+        Element tilesetSourceElement = (Element) tilesetSource.item(0);
+
+        NodeList tileset = loadXmlTileset("./map/" + tilesetSourceElement.getAttribute("source"));
```
Comment (RouNNdeL) [inline core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java]: The path here should be handled better, ex. get the folder from the `path` variable (use `File`)

File: core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+        File inputFile = new File(path);
+        DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
+        DocumentBuilder documentBuilder = documentBuilderFactory.newDocumentBuilder();
+        Document document = documentBuilder.parse(inputFile);
+        document.getDocumentElement().normalize();
+        return document.getDocumentElement();
+    }
+
+    private static NodeList loadXmlTileset(String path) throws IOException, ParserConfigurationException, SAXException {
```
Comment (RouNNdeL) [inline core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java]: Same as in `loadXmlMap`

File: core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                    Gdx.app.error(TAG, "Unsupported tile layer found: '" +
+                        element.getAttribute("name") + "', will be ignored");
+            }
+        }
+        return new TileMap(baseLayer, floorLayer, wallsLayer);
+    }
+
+    private static Node loadXmlMap(String path) throws IOException, ParserConfigurationException, SAXException {
+
+        File inputFile = new File(path);
```
Comment (RouNNdeL) [inline core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java]: Better to pass `File` as the argument here.

```suggestion
    private static Node loadXmlMap(File file) throws IOException, ParserConfigurationException, SAXException {

        // Rename inputFile to file
        // File inputFile = new File(path);
```

File: core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    private static Tile[][] csvStringToTileArray(String input, int width, int height, NodeList tileset, World world)
+        throws InvalidPropertiesFormatException {
+        String[] mapStringArray = input.trim().replaceAll("\n", "").split(",");
+        Tile[][] tiles = new Tile[height][width];
+        for (int y = 0; y < height; y++) {
+            for (int x = 0; x < width; x++) {
+                tiles[x][y] = TileFactory.createTile(getTileById(tileset, mapStringArray[(height - y - 1) * width + x]),
+                    world,
+                    x,
+                    y);
```
Comment (RouNNdeL) [inline core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java]: Can be put on the same line

File: core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+                    break;
+                case LAYER_WALLS:
+                    wallsLayer = mapLayer;
+                    break;
+                default:
+                    Gdx.app.error(TAG, "Unsupported tile layer found: '" +
+                        element.getAttribute("name") + "', will be ignored");
+            }
+        }
+        return new TileMap(baseLayer, floorLayer, wallsLayer);
```
Comment (RouNNdeL) [inline core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java:81]: Blank lines before `return` statements are recommended,

File: core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        NodeList tilesetSource = mapElement.getElementsByTagName("tileset");
+        Element tilesetSourceElement = (Element) tilesetSource.item(0);
+
+        NodeList tileset = loadXmlTileset("./map/" + tilesetSourceElement.getAttribute("source"));
+
+        TileMapLayer baseLayer = null;
+        TileMapLayer floorLayer = null;
+        TileMapLayer wallsLayer = null;
+
+
```
Comment (RouNNdeL) [inline core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java]: Multiple blank lines are discouraged.

File: core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+public class TileMapFactory {
+    private static final String TAG = TileMapFactory.class.getSimpleName();
+
+    private static final String LAYER_BASE = "base";
+    private static final String LAYER_FLOOR = "floor";
+    private static final String LAYER_WALLS = "walls";
+
+
+    public static TileMap createTileMap(World world, String path) throws IOException, MissingLayersException, ParserConfigurationException, SAXException {
+
```
Comment (RouNNdeL) [inline core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java]: First line in a function or method should not be empty.

File: core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+import java.util.InvalidPropertiesFormatException;
+
+public class TileMapFactory {
+    private static final String TAG = TileMapFactory.class.getSimpleName();
+
+    private static final String LAYER_BASE = "base";
+    private static final String LAYER_FLOOR = "floor";
+    private static final String LAYER_WALLS = "walls";
+
+
```
Comment (RouNNdeL) [inline core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java]: Multiple blank lines are discouraged.

File: core/src/main/java/com/cyberbot/bomberman/core/models/tiles/TileFactory.java [hunk 9]
```diff
@@ -1,55 +1,71 @@
...
-                    Tile.PROPERTY_TEXTURE_NAME + "' properties"
+                    PROPERTY_TILE_TYPE + "' and '" +
+                    PROPERTY_TEXTURE_NAME + "' properties"
             );
         }
 
-        String tileType = properties.get(Tile.PROPERTY_TILE_TYPE, String.class);
-        String textureName = properties.get(Tile.PROPERTY_TEXTURE_NAME, String.class);
+        String tileType = (String) properties.get(PROPERTY_TILE_TYPE);
+        String textureName = (String) properties.get(PROPERTY_TEXTURE_NAME);
```
Comment (RouNNdeL) [inline core/src/main/java/com/cyberbot/bomberman/core/models/tiles/TileFactory.java]: Should catch `ClassCastException` and throw `InvalidPropertiesFormatException`

File: core/build.gradle [hunk 10]
```diff
@@ -16,4 +16,5 @@ dependencies {
     testImplementation 'org.junit.jupiter:junit-jupiter-params:5.6.0'
 
     testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.6.0'
+
```
Comment (RouNNdeL) [inline core/build.gradle]: Unnecessary file change, please revert.

File: client/core/src/com/cyberbot/bomberman/screens/GameScreen.java [hunk 11]
```diff
@@ -18,10 +18,14 @@
 import com.cyberbot.bomberman.core.models.entities.PlayerEntity;
 import com.cyberbot.bomberman.core.models.tiles.MissingLayersException;
 import com.cyberbot.bomberman.core.models.tiles.TileMap;
+import com.cyberbot.bomberman.core.models.tiles.loader.TileMapFactory;
 import com.cyberbot.bomberman.models.KeyBinds;
+import org.xml.sax.SAXException;
 
+import javax.xml.bind.JAXBException;
```
Comment (RouNNdeL) [inline client/core/src/com/cyberbot/bomberman/screens/GameScreen.java]: Invalid import.

File: .idea/libraries/Java_EE_6_Java_EE_6.xml [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+      <root url="jar://$PROJECT_DIR$/lib/javax.ejb.jar!/" />
+      <root url="jar://$PROJECT_DIR$/lib/javax.resource.jar!/" />
+      <root url="jar://$PROJECT_DIR$/lib/javax.jms.jar!/" />
+      <root url="jar://$PROJECT_DIR$/lib/javax.servlet.jar!/" />
+      <root url="jar://$PROJECT_DIR$/lib/javax.servlet.jsp.jstl.jar!/" />
+    </CLASSES>
+    <JAVADOC />
+    <SOURCES />
+  </library>
+</component>
```
Comment (RouNNdeL) [inline .idea/libraries/Java_EE_6_Java_EE_6.xml]: Is this library needed?

File: client/core/src/com/cyberbot/bomberman/managers/GameScreenManager.java [hunk 13]
```diff
@@ -4,9 +4,13 @@
 import com.cyberbot.bomberman.core.models.tiles.MissingLayersException;
 import com.cyberbot.bomberman.screens.AbstractScreen;
 import com.cyberbot.bomberman.screens.GameScreen;
+import org.xml.sax.SAXException;
 
+import javax.xml.bind.JAXBException;
```
Comment (RouNNdeL) [inline client/core/src/com/cyberbot/bomberman/managers/GameScreenManager.java]: Invalid import.

**Review by RouNNdeL — CHANGES_REQUESTED** [review]
Please address this minor comments and add some basic documentation to the public API.

File: client/core/src/com/cyberbot/bomberman/managers/GameScreenManager.java (hunk 13 above)
Comment (BWalusiak) [reply client/core/src/com/cyberbot/bomberman/managers/GameScreenManager.java]: Removed

File: .idea/libraries/Java_EE_6_Java_EE_6.xml (hunk 12 above)
Comment (BWalusiak) [reply .idea/libraries/Java_EE_6_Java_EE_6.xml]: Removed

File: client/core/src/com/cyberbot/bomberman/screens/GameScreen.java (hunk 11 above)
Comment (BWalusiak) [reply client/core/src/com/cyberbot/bomberman/screens/GameScreen.java]: Removed

File: core/build.gradle (hunk 10 above)
Comment (BWalusiak) [reply core/build.gradle]: Reverted

File: core/src/main/java/com/cyberbot/bomberman/core/models/tiles/TileFactory.java (hunk 9 above)
Comment (BWalusiak) [reply core/src/main/java/com/cyberbot/bomberman/core/models/tiles/TileFactory.java]: Fixed

File: core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java (hunk 8 above)
Comment (BWalusiak) [reply core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java]: Removed

File: core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java (hunk 7 above)
Comment (BWalusiak) [reply core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java]: Removed

File: core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java (hunk 6 above)
Comment (BWalusiak) [reply core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java]: Removed

File: core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java (hunk 5 above)
Comment (BWalusiak) [reply core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java:81]: Fixed

File: core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java (hunk 4 above)
Comment (BWalusiak) [reply core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java]: Fixed

File: core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java (hunk 3 above)
Comment (BWalusiak) [reply core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java]: Fixed

File: core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java (hunk 2 above)
Comment (BWalusiak) [reply core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java]: Fixed

File: core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java (hunk 1 above)
Comment (BWalusiak) [reply core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java]: Fixed

File: core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+     * @throws SAXException                 {@link DocumentBuilder DOM parser} exception.
+     */
+    public static TileMap createTileMap(World world, String path) throws IOException,
+        MissingLayersException, ParserConfigurationException, SAXException {
+        File mapFile = new File(path);
+
+        Element sourceMap = (Element) loadXmlMap(mapFile);
+        Element tilesetSourceElement = (Element) sourceMap.getElementsByTagName("tileset").item(0);
+
+        File tilesetFile = new File(mapFile.getParent() + "/" + tilesetSourceElement.getAttribute("source"));
```
Comment (RouNNdeL) [inline core/src/main/java/com/cyberbot/bomberman/core/models/tiles/loader/TileMapFactory.java]:
```suggestion
        File tilesetFile = new File(mapFile.getParent() + File.separator + tilesetSourceElement.getAttribute("source"));
```

**Review by RouNNdeL — CHANGES_REQUESTED** [review]

**Review by RouNNdeL — APPROVED** [review]
Good job!

**Pull request merged**